src/audio/mux/mux.c

```diff
@@ -610,48 +610,116 @@ static int mux_copy(struct comp_dev *dev)
 
 static int mux_reset(struct comp_dev *dev)
 {
+	int dir = dev->pipeline->source_comp->direction;
+	struct list_item *blist;
+	struct comp_buffer *source;
+
 	comp_info(dev, "mux_reset()");
 
-	return comp_set_state(dev, COMP_TRIGGER_RESET);
+	if (dir == SOF_IPC_STREAM_PLAYBACK) {
+		list_for_item(blist, &dev->bsource_list) {
+			source = container_of(blist, struct comp_buffer,
+					      sink_list);
+			/* only mux the sources with the same state with mux */
+			if (source->source->state > COMP_STATE_READY)
+				/* should not reset the downstream components */
+				return 1;
+		}
+	}
+
+	comp_set_state(dev, COMP_TRIGGER_RESET);
+	return 0;
 }
 
 static int mux_prepare(struct comp_dev *dev)
 {
 	struct comp_data *cd = comp_get_drvdata(dev);
+	struct list_item *blist;
+	struct comp_buffer *source;
+	int downstream = 0;
 	int ret;
 
 	comp_info(dev, "mux_prepare()");
 
-	if (dev->comp.type == SOF_COMP_MUX)
-		cd->mux = mux_get_processing_function(dev);
-	else
-		cd->demux = demux_get_processing_function(dev);
+	if (dev->state != COMP_STATE_ACTIVE) {
+		if (dev->comp.type == SOF_COMP_MUX)
+			cd->mux = mux_get_processing_function(dev);
+		else
+			cd->demux = demux_get_processing_function(dev);
 
-	if (!cd->mux && !cd->demux) {
-		comp_err(dev, "mux_prepare(): Invalid configuration, couldn't find suitable processing function.");
-		return -EINVAL;
+		if (!cd->mux && !cd->demux) {
+			comp_err(dev, "mux_prepare(): Invalid configuration, couldn't find suitable processing function.");
+			return -EINVAL;
+		}
+
+		ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+		if (ret) {
+			comp_info(dev, "mux_prepare() comp_set_state() returned non-zero.");
+			return ret;
+		}
 	}
 
-	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
-	if (ret) {
-		comp_info(dev, "mux_prepare() comp_set_state() returned non-zero.");
-		return ret;
+	/* check each mixer source state */
+	list_for_item(blist, &dev->bsource_list) {
```

Comment:
Should this be mux instead of mixer?

Reply:
yes will fix it. I copied this directly from mixer....